test/src/test/java/hudson/model/UserTest.java

```diff
@@ -783,4 +773,186 @@ public WebResponse getResponse(WebRequest request) throws IOException {
+                            return false;
+                        }
+                    };
+
+                case "expired":
+                    return new UserStub(username) {
+                        @Override
+                        public boolean isAccountNonExpired() {
+                            return false;
+                        }
+                    };
+
+                case "password_expired":
+                    return new UserStub(username) {
+                        @Override
+                        public boolean isCredentialsNonExpired() {
+                            return false;
+                        }
+                    };
+
+                default:
+                    return new UserStub(username);
+            }
+        }
+
+        @Override
+        public GroupDetails loadGroupByGroupname(String groupname) throws UsernameNotFoundException, DataAccessException {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @Test
+    @Issue("JENKINS-55813")
+    public void lockedUser() throws Exception {
+        checkUserImpersonationOperations("locked", "user is locked");
+    }
+
+    @Test
+    @Issue("JENKINS-55813")
+    public void disabledUser() throws Exception {
+        checkUserImpersonationOperations("disabled", "user is disabled");
+    }
+
+    @Test
+    @Issue("JENKINS-55813")
+    public void expiredUser() throws Exception {
+        checkUserImpersonationOperations("expired", "user expired or inactive");
+    }
+
+    @Test
+    @Issue("JENKINS-55813")
+    public void passwordExpired() throws Exception {
+        checkUserImpersonationOperations("password_expired", "user credentials expired");
+    }
+
+    @Test
+    @Issue("JENKINS-55813")
+    public void enabledUser() throws Exception {
+        j.jenkins.setSecurityRealm(new DisabledAccountsRealm());
+        User user = User.getOrCreateByIdOrFullName("enabled");
+        assertNotNull("Expected user impersonation to work for enabled users",
+                user.impersonate());
+        assertNotNull("Expected login to work for enabled users",
+                j.createWebClient().login("enabled"));
+        assertNotNull("Expected API tokens to work for enabled users",
+                j.createWebClient().withBasicApiToken(user).getPage(j.jenkins));
+
+        SSHD sshd = SSHD.get();
+        sshd.setPort(0);
+        sshd.start();
+        try (SshClient client = SshClient.setUpDefaultClient()) {
+            client.setServerKeyVerifier(AcceptAllServerKeyVerifier.INSTANCE);
+            client.start();
+            ConnectFuture future = client.connect("enabled", new InetSocketAddress(sshd.getActualPort()));
+            try (ClientSession session = future.verify(10, TimeUnit.SECONDS).getSession()) {
+                session.addPublicKeyIdentity(generateUserSshKey(user));
+                assertTrue("Expected successful SSH authentication", session.auth().await(10, TimeUnit.SECONDS));
```

Comment:
It seems you are really testing behaviors of `ssh-cli-auth` and/or `sshd` here? Best to remove this part of the test, or leave it as an integration test in one of those modules, which will also be cleaner when I resurrect #3988 in simplified form.

Reply:
This test was added at @jeffret-b's request since the underlying issue related to API tokens and SSH keys. However, if I migrate this functionality all to the LDAP plugin specifically, then I could move these tests there.

Reply:
Or to `sshd-module`.

Reply:
Then that requires `sshd-module` to depend on a version of Jenkins that contains this patch?

Reply:
I've added a comment in these two methods pointing out that the tests are common authentication mechanisms (you can see web login, API token login, and SSH login).

Reply:
> Then that requires `sshd-module` to depend on a version of Jenkins that contains this patch?

Yes, which you can do temporarily using an incremental version, then switch to a weekly if and when this is released.

(Remember that there is no reason to hold back the `jenkins.version` in a `jenkins-module` since they are not distributed to users via the UC.)

Reply:
API tokens are implemented in core so the test for it needs to be in core. This is probably as reasonable a place for it as anywhere. Putting the ssh test in sshd-module would be a better location.

Reply:
I was not aware about that regarding modules. Now that I'm a maintainer of sshd-module, that's not even too difficult to move this to. Good idea.

Reply:
Tests moved to https://github.com/jenkinsci/sshd-module/pull/39 which also discovered a bug in this test as it is.